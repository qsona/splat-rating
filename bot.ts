// Require the necessary discord.js classes
import { Client, GatewayIntentBits, CommandInteraction, InteractionType, Interaction, ChatInputCommandInteraction } from 'discord.js'

import registerHandler from './src/commands/register'
import newgameHandler from './src/commands/newgame'
import joinHandler from './src/commands/join'
import leaveHandler from './src/commands/leave'
import breakHandler from './src/commands/break'
import matchHandler from './src/commands/match'
import resultHandler from './src/commands/report'
import makeRegisterHandler from './src/commands/make-register'
import makeJoinHandler from './src/commands/make-join'
import makeLeaveHandler from './src/commands/make-leave'
import displayHandler from './src/commands/display'
import pingHandler from './src/commands/ping'
import separateHandler from './src/commands/separate'
import separationResetHandler from './src/commands/separation-reset'

import { tksPartyHandler, tksRecruitHandler } from './src/commands/tks'

import { execute as executeButtonHandlers } from './src/commands/buttonHandlers'
import { execute as executeModalHandlers } from './src/commands/modalHandlers'

require('dotenv').config()

const token = process.env.DISCORD_TOKEN

// Create a new client instance
const client = new Client({ intents: [GatewayIntentBits.Guilds] })

// When the client is ready, run this code (only once)
client.once('ready', () => {
  console.log('Ready!')
})

// Login to Discord with your client's token
if (token) {
  client.login(token)
} else {
  console.log('DISCORD_TOKEN is not set')
}

export interface CommandHandler {
  commandName: string
  execute: (interaction: ChatInputCommandInteraction) => Promise<void>
}

const handlers = new Map<string, CommandHandler>()

const helpHandler: CommandHandler = {
  commandName: 'sr-help',
  execute: async (interaction) => {
    await interaction.reply('参加する人は /sr-register でレーティング登録をして /sr-join で参加してね! \n参考URL: https://qsona.github.io/splat-rating/')
  },
}

;[
  registerHandler,
  newgameHandler,
  joinHandler,
  leaveHandler,
  matchHandler,
  breakHandler,
  resultHandler,
  makeRegisterHandler,
  makeJoinHandler,
  makeLeaveHandler,
  displayHandler,
  pingHandler,
  helpHandler,
  separateHandler,
  separationResetHandler,
  tksRecruitHandler,
  tksPartyHandler,
].forEach((handler) => {
  handlers.set(handler.commandName, handler)
})

client.on('interactionCreate', async (interaction) => {
  // if (interaction.type !== InteractionType.ApplicationCommand) {
  if (interaction.isChatInputCommand()) {
    const { commandName } = interaction
    console.log(commandName)
    console.log(interaction.user)

    const handler = handlers.get(commandName)
    if (handler) {
      await handler.execute(interaction)
    }
    return
  }

  if (interaction.isButton()) {
    await executeButtonHandlers(interaction)
    return
  }

  if (interaction.type === InteractionType.ModalSubmit) {
    await executeModalHandlers(interaction)
    return
  }
})
