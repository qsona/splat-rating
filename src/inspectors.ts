import { TksRating, TksTeam } from '@prisma/client'
import { sumBy } from 'lodash'
import { tksTeamRatingInfo } from './models/TksTeam'

export const inspectR = (r: number) => `R${Math.floor(r)}`

export const inspectTeamUsers = (teamUsers: { mu: number; username: string }[]) => {
  //  const usersStr = teamUsers.map((ru) => `${ru.username} (R${ru.mu})`).join(' ')
  const usersStr = teamUsers.map((ru) => ru.username).join(' ')
  const ratingStr = `合計${inspectR(sumBy(teamUsers, (ru) => ru.mu))} (${teamUsers.map((tu) => inspectR(tu.mu)).join(', ')})`

  return `${usersStr} | ${ratingStr}`
}

export const inspectTksTeam = (team: TksTeam, rating: TksRating) => {
  const { rank, rate, isTentative } = tksTeamRatingInfo(team, rating)
  const rateText = isTentative ? `[計測中 ${rating.winCount}/5] (推定${inspectR(rate)})` : `${rank.iconText} (${inspectR(rate)})`
  return `${rateText} | ${team.name || '(チーム名未設定)'}`
}
