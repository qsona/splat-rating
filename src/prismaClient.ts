import { PrismaClient } from '@prisma/client'

export const prisma = new PrismaClient({ log: ['info', 'query', 'warn', 'error'] })
