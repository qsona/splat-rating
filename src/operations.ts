import assert from 'assert'
import { sum, difference } from 'lodash'
import Aigle from 'aigle'
import { JoinableGame, Rating, JoinedUser, GameMatching, PlayingGame } from '@prisma/client'
import { prisma } from './prismaClient'
import { rate } from 'openskill'
import { SplatRuleSet } from './rules'

const BETA = 200
const INITIAL_SIGMA = 200

export const createNewGame = async (userId: string, discordChannelId: string, rule: SplatRuleSet, totalGameCount: number) => {
  const rating = await prisma.rating.findUnique({
    where: {
      userId_rule: { userId, rule },
    },
  })
  if (!rating) {
    return 'RATING_DOES_NOT_EXIST'
  }

  return await prisma.$transaction(async (prisma) => {
    let joinableGame: JoinableGame
    try {
      joinableGame = await prisma.joinableGame.create({
        data: {
          discordChannelId,
          creatorUserId: userId,
          rule,
          totalGameCount,
        },
      })
    } catch (e) {
      if ((e as any).code === 'P2002') {
        return 'JOINABLE_GAME_ALREADY_EXISTS'
      }
      throw e
    }
    const joinedUser = await prisma.joinedUser.create({
      data: {
        joinableGameId: joinableGame.id,
        userId,
      },
    })

    return { rating, joinableGame, joinedUser }
  })
}

export const joinGame = async (userId: string, discordChannelId: string) => {
  // TODO: acquire lock
  const joinableGame = await prisma.joinableGame.findUnique({
    where: { discordChannelId },
  })
  if (!joinableGame) {
    return 'JOINABLE_GAME_DOES_NOT_EXIST'
  }
  const { rule } = joinableGame
  const rating = await prisma.rating.findUnique({
    where: {
      userId_rule: { userId, rule },
    },
  })
  if (!rating) {
    return 'RATING_DOES_NOT_EXIST'
  }

  return await prisma.$transaction(async (prisma) => {
    let joinedUser: JoinedUser
    try {
      joinedUser = await prisma.joinedUser.create({
        data: {
          joinableGameId: joinableGame.id,
          userId,
        },
      })
    } catch (e) {
      if ((e as any).code === 'P2002') {
        return 'USER_ALREADY_JOINED'
      }
      throw e
    }

    const joinedUsers = await prisma.joinedUser.findMany({
      where: {
        joinableGameId: joinableGame.id,
      },
    })

    if (joinedUsers.length < 8) {
      return {
        status: 'joinable' as 'joinable',
        joinedUser,
        joinableGame,
        joinedUsersCount: joinedUsers.length,
        remainUsersCount: 8 - joinedUsers.length,
        rating,
      }
    }

    // matching
    assert.equal(joinedUsers.length, 8)

    const ratings = await prisma.rating.findMany({
      where: {
        userId: { in: joinedUsers.map((u) => u.userId) },
        rule,
      },
    })
    if (ratings.length !== joinedUsers.length) {
      throw new Error(`mismatch between ratings and joinedUsers`)
    }

    const { creatorUserId } = joinableGame

    // move creator rating to first
    const creatorRatingIndex = ratings.findIndex((r) => r.userId === creatorUserId)
    if (creatorRatingIndex < 0) {
      throw new Error(`rating of creator does not exist`)
    }
    const creatorRating = ratings[creatorRatingIndex]
    ratings.splice(creatorRatingIndex, 1)
    ratings.unshift(creatorRating)

    const playingGame = await prisma.playingGame.create({
      data: {
        id: joinableGame.id,
        discordChannelId,
        rule,
        creatorUserId,
        totalGameCount: joinableGame.totalGameCount,
        currentGameCount: 0,
      },
    })

    const gameMatching = await createGameMatching(playingGame.id, ratings)

    await prisma.joinedUser.deleteMany({ where: { joinableGameId: joinableGame.id } })
    await prisma.joinableGame.delete({ where: { id: joinableGame.id } })

    return {
      status: 'matched' as 'matched',
      playingGame,
      rating,
      gameMatching,
    }
  })
}

export const createRating = async (userId: string, rule: SplatRuleSet, estimatedGachiPower: number) => {
  const rating = await prisma.rating.create({
    data: {
      mu: estimatedGachiPower,
      sigma: INITIAL_SIGMA,
      userId: userId,
      rule,
    },
  })

  return rating
}

export const createGameMatching = async (playingGameId: string, ratings: Rating[]) => {
  assert.equal(ratings.length, 8)
  const teamsRatings = calculateMatchingWithMinRateDiff(ratings)

  const gameMatching = await prisma.gameMatching.create({
    data: {
      playingGameId,
      teamsRatingIds: teamsRatings.map((teamRatings) => teamRatings.map((r) => r.id)),
    },
  })
  return gameMatching
}

export const calculateMatchingWithMinRateDiff = (ratings: Rating[]) => {
  let minRateDiff = Number.POSITIVE_INFINITY
  let rates = ratings.map((r) => r.mu)
  let totalRate = sum(rates)
  let currentAlphaTeamIndexes: number[] = []

  for (let i = 1; i <= 5; i++) {
    for (let j = i + 1; j <= 6; j++) {
      for (let k = j + 1; k <= 7; k++) {
        const alphaTeamIndexes = [0, i, j, k]
        const alphaTeamTotalRate = sum(alphaTeamIndexes.map((i) => rates[i]))
        const bravoTeamTotalRate = totalRate - alphaTeamTotalRate
        const rateDiff = Math.abs(alphaTeamTotalRate - bravoTeamTotalRate)

        if (rateDiff < minRateDiff) {
          currentAlphaTeamIndexes = alphaTeamIndexes
          minRateDiff = rateDiff
        }
      }
    }
  }

  const decidedAlphaTeamIndexes = currentAlphaTeamIndexes
  const decidedBravoTeamIndexes = difference([0, 1, 2, 3, 4, 5, 6, 7], decidedAlphaTeamIndexes)
  return [decidedAlphaTeamIndexes, decidedBravoTeamIndexes].map((teamIndexes) => {
    return teamIndexes.map((index) => ratings[index])
  })
}

export type RatingResult = {
  ratingId: string
  before: { mu: number; sigma: number }
  after: { mu: number; sigma: number }
}

export const createGameResult = async (playingGame: PlayingGame, isAlphaWin: boolean) => {
  const gameMatching = await prisma.gameMatching.findUnique({ where: { playingGameId: playingGame.id }, rejectOnNotFound: true })

  return await prisma.$transaction(async (prisma) => {
    const teamsRatings = await Aigle.map(gameMatching.teamsRatingIds as string[][], async (teamRatingIds) => {
      const teamRatings = await prisma.rating.findMany({ where: { id: { in: teamRatingIds } } })
      return teamRatingIds.map((teamRatingId) => {
        const rating = teamRatings.find((r) => r.id === teamRatingId)
        if (!rating) {
          throw new Error(`rating not found. id: ${teamRatingId}`)
        }
        return rating
      })
    })
    const sortedTeamsRatings = isAlphaWin ? [teamsRatings[0], teamsRatings[1]] : [teamsRatings[1], teamsRatings[0]]
    const newTeamsRatings = rate(sortedTeamsRatings, { beta: BETA })
    console.log('win', newTeamsRatings[0])
    console.log('lose', newTeamsRatings[1])
    const [winnerTeamsRatings, loserTeamsRatings]: RatingResult[][] = await Aigle.map(sortedTeamsRatings, async (teamRatings, teamIndex) => {
      return await Aigle.map(teamRatings, async (rating, ratingIndex) => {
        const newRating = newTeamsRatings[teamIndex][ratingIndex]
        await prisma.rating.update({
          where: {
            id: rating.id,
          },
          data: {
            mu: newRating.mu,
            sigma: newRating.sigma,
          },
        })

        return {
          ratingId: rating.id,
          before: {
            mu: rating.mu,
            sigma: rating.sigma,
          },
          after: {
            mu: newRating.mu,
            sigma: newRating.sigma,
          },
        }
      })
    })

    const gameResult = await prisma.gameResult.create({
      data: {
        beta: BETA,
        winnerTeamsRatings,
        loserTeamsRatings,
      },
    })

    await prisma.gameMatching.delete({
      where: { id: gameMatching.id },
    })

    const isFinishing = playingGame.currentGameCount + 1 === playingGame.totalGameCount

    if (isFinishing) {
      await prisma.playingGame.delete({
        where: { id: playingGame.id },
      })
      return {
        isFinished: true as true,
        gameResult,
      }
    }

    const previousRatingIds = gameMatching.teamsRatingIds as string[][]

    // create next match
    // TODO
    // const nextGameMatching = await createGameMatching(playingGame.id, [])

    return {
      isFinished: false as false,
      gameResult,
      // nextGameMatching,
    }
  })
}
