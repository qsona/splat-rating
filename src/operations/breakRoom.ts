import { prisma } from '../prismaClient'

export const breakRoom = async (userId: string, discordChannelId: string) => {
  // TODO: acquire lock
  const room = await prisma.room.findUnique({
    where: { discordChannelId },
  })
  if (!room) {
    return 'ROOM_DOES_NOT_EXIST'
  }

  // TODO: who can break room?

  const matchingExists = !!(await prisma.matching.count({
    where: {
      roomId: room.id,
    },
  }))

  if (matchingExists) {
    return 'MATCHING_EXISTS'
  }

  await prisma.$transaction(async (prisma) => {
    await prisma.joinedUser.deleteMany({ where: { roomId: room.id } })
    await prisma.room.delete({ where: { id: room.id } })
  })
  return { success: true }
}
