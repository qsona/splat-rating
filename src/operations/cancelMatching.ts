import { prisma } from '../prismaClient'

export const cancelMatching = async (discordChannelId: string) => {
  const room = await prisma.room.findUnique({
    where: { discordChannelId },
  })
  if (!room) {
    return 'ROOM_DOES_NOT_EXIST'
  }

  const matching = await prisma.matching.findUnique({ where: { roomId: room.id } })
  if (!matching) {
    return 'MATCHING_DOES_NOT_EXIST'
  }

  await prisma.matching.delete({
    where: { id: matching.id },
  })

  return { success: true }
}
