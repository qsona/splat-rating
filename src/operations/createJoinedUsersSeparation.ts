import { intersection } from 'lodash'
import { prisma } from '../prismaClient'

export const createJoinedUsersSeparation = async (userIds: [string, string], discordChannelId: string) => {
  if (userIds[0] === userIds[1]) {
    return { error: 'SAME_USER_SPECIFIED' as const }
  }

  const room = await prisma.room.findUnique({
    where: { discordChannelId },
    include: { joinedUsers: true, joinedUsersSeparations: true },
  })
  if (!room) {
    return { error: 'ROOM_DOES_NOT_EXIST' as const }
  }

  const separatingJoinedUsers = room.joinedUsers.filter((ju) => userIds.includes(ju.userId))
  if (separatingJoinedUsers.length !== 2) {
    return { error: 'USER_NOT_JOINED' as const }
  }

  const joinedUsersSeparationUserIds = room.joinedUsersSeparations.flatMap((jus) => [jus.firstJoinedUserId, jus.secondJoinedUserId])
  const conflictedUserIds = intersection(joinedUsersSeparationUserIds, userIds)
  if (conflictedUserIds.length) {
    return { error: 'SEPARATION_CONFLICTED' as const, conflictedUserIds }
  }

  const joinedUsersSeparation = await prisma.joinedUsersSeparation.create({
    data: {
      firstJoinedUserId: separatingJoinedUsers[0].id,
      secondJoinedUserId: separatingJoinedUsers[1].id,
      roomId: room.id,
    },
  })

  return {
    joinedUsersSeparation,
  }
}
