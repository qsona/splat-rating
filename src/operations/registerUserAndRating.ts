import { SplatRuleSet } from '../rules'
import { prisma } from '../prismaClient'
import { Rating, User } from '@prisma/client'

export const INITIAL_SIGMA = 200

export const registerUserAndRating = async (userId: string, username: string, guildId: string, rule: SplatRuleSet, estimatedGachiPower: number) => {
  let isNewUser = true
  let user: User | null = await prisma.user.findUnique({
    where: {
      id: userId,
    },
  })
  if (user) {
    isNewUser = false
  } else {
    try {
      user = await prisma.user.create({
        data: {
          id: userId,
          name: username,
        },
      })
    } catch (e) {
      if ((e as any).code === 'P2002') {
        // just user is already registered
        isNewUser = false
      } else {
        throw e
      }
    }
  }

  let rating: Rating | null = await prisma.rating.findUnique({
    where: {
      userId_guildId_rule: {
        userId,
        guildId,
        rule,
      },
    },
  })
  if (rating) {
    return 'RATING_ALREADY_REGISTERED' as const
  }

  try {
    rating = await prisma.rating.create({
      data: {
        guildId,
        mu: estimatedGachiPower,
        sigma: INITIAL_SIGMA,
        user: { connect: { id: userId } },
        rule,
      },
    })
  } catch (e) {
    if ((e as any).code === 'P2002') {
      return 'RATING_ALREADY_REGISTERED' as const
    }
    throw e
  }

  if (isNewUser) {
    return {
      isNewUser: true as true,
      user: user!,
      rating,
    }
  } else {
    return {
      isNewUser: false as false,
      rating,
    }
  }
}
