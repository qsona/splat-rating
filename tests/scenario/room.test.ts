import { it } from '@jest/globals'
import { User } from '@prisma/client'
import assert from 'assert'
import { onInteractionCreated } from '../../bot'
import { ButtonBuilder, InteractionReplyOptions } from 'discord.js'
import { prisma } from '../../src/prismaClient'
import { registerUserAndRating } from '../../src/operations/registerUserAndRating'

const guildId = 'guild1'
const channelId = 'channel1'

const getCustomId = (component: any) => component?.toJSON().components?.[0]?.custom_id

const createChatInputInteraction = (command: string, args: any, onReply: (options: InteractionReplyOptions) => void) => {
  return {
    isChatInputCommand: () => true,
    isButton: () => false,
    commandName: command,
    reply: (arg: any) => onReply(typeof arg === 'string' ? { content: arg } : arg),
    ...args,
  }
}

const createButtonInteraction = (command: string, args: any, onReply: (options: InteractionReplyOptions) => void) => {
  return {
    isChatInputCommand: () => false,
    isButton: () => true,
    customId: command,
    reply: (arg: any) => onReply(typeof arg === 'string' ? { content: arg } : arg),
    ...args,
  }
}

const createButtonWithDataInteraction = (command: string, data: string, args: any, onReply: (options: InteractionReplyOptions) => void) => {
  return {
    isChatInputCommand: () => false,
    isButton: () => true,
    customId: `command@data`,
    reply: (arg: any) => onReply(typeof arg === 'string' ? { content: arg } : arg),
    ...args,
  }
}

describe('room scenario', () => {
  it('room scenario', async () => {
    // prepare
    const user1 = await registerUserAndRating('user1', 'username1', guildId, 'SplatZones', 2000)
    const user2 = await registerUserAndRating('user2', 'username2', guildId, 'SplatZones', 2100)
    const user3 = await registerUserAndRating('user3', 'username3', guildId, 'SplatZones', 2200)
    const user4 = await registerUserAndRating('user4', 'username4', guildId, 'SplatZones', 2300)
    const user5 = await registerUserAndRating('user5', 'username5', guildId, 'SplatZones', 2400)
    const user6 = await registerUserAndRating('user6', 'username6', guildId, 'SplatZones', 2500)
    const user7 = await registerUserAndRating('user7', 'username7', guildId, 'SplatZones', 2600)
    const user8 = await registerUserAndRating('user8', 'username8', guildId, 'SplatZones', 2700)
    const user9 = await registerUserAndRating('user9', 'username9', guildId, 'SplatZones', 2800)
    const user10 = await registerUserAndRating('user10', 'username10', guildId, 'SplatZones', 2900)

    // new game
    await onInteractionCreated(
      createChatInputInteraction(
        'sr-newgame',
        {
          channelId,
          guildId,
          user: {
            id: 'user1',
            username: 'username1',
          },
          options: {
            getString: () => 'SplatZones', // TODO: use mock
          },
        },
        (options) => {
          expect(options.components).toHaveLength(2)
          expect(getCustomId(options.components![0])).toBe('button-join')
          expect(getCustomId(options.components![1])).toBe('button-leave')

          expect(options).toMatchSnapshot()
        }
      )
    )

    const room = await prisma.room.findFirst({
      where: {
        creatorUserId: 'user1',
        discordChannelId: channelId,
      },
    })
    expect(room).toBeTruthy()

    // join
    await onInteractionCreated(
      createButtonInteraction(
        'button-join',
        {
          channelId,
          guildId,
          user: {
            id: 'user2',
            username: 'username2',
          },
        },
        (options) => {
          expect(options.components).toHaveLength(2)
          expect(getCustomId(options.components![0])).toBe('button-join')
          expect(getCustomId(options.components![1])).toBe('button-leave')

          expect(options).toMatchSnapshot()
        }
      )
    )

    // leave
    await onInteractionCreated(
      createButtonInteraction(
        'button-leave',
        {
          channelId,
          guildId,
          user: {
            id: 'user2',
            username: 'username2',
          },
        },
        (options) => {
          expect(options).toMatchSnapshot()
        }
      )
    )

    // join 2-7
    for (let i = 2; i <= 7; i++) {
      await onInteractionCreated(
        createButtonInteraction(
          'button-join',
          {
            channelId,
            guildId,
            user: {
              id: `user${i}`,
              username: `username${i}`,
            },
          },
          (options) => {
            expect(options.components).toHaveLength(2)
            expect(getCustomId(options.components![0])).toBe('button-join')
            expect(getCustomId(options.components![1])).toBe('button-leave')
          }
        )
      )
    }
    // join 8
    await onInteractionCreated(
      createButtonInteraction(
        'button-join',
        {
          channelId,
          guildId,
          user: {
            id: 'user8',
            username: 'username8',
          },
        },
        (options) => {
          expect(options.components).toHaveLength(3)
          expect(getCustomId(options.components![0])).toBe('button-match')
          expect(getCustomId(options.components![1])).toBe('button-join')
          expect(getCustomId(options.components![2])).toBe('button-leave')
          expect(options).toMatchSnapshot()
        }
      )
    )

    // leave
    // 7人になる
    await onInteractionCreated(
      createButtonInteraction(
        'button-leave',
        {
          channelId,
          guildId,
          user: {
            id: 'user2',
            username: 'username2',
          },
        },
        (options) => {
          // no button-match
          expect(options.components || []).toHaveLength(0)
          expect(options).toMatchSnapshot()
        }
      )
    )

    // re-join 8
    await onInteractionCreated(
      createButtonInteraction(
        'button-join',
        {
          channelId,
          guildId,
          user: {
            id: 'user2',
            username: 'username2',
          },
        },
        (options) => {
          expect(options.components).toHaveLength(3)
          expect(getCustomId(options.components![0])).toBe('button-match')
          expect(getCustomId(options.components![1])).toBe('button-join')
          expect(getCustomId(options.components![2])).toBe('button-leave')
        }
      )
    )

    // join 9
    await onInteractionCreated(
      createButtonInteraction(
        'button-join',
        {
          channelId,
          guildId,
          user: {
            id: 'user9',
            username: 'username9',
          },
        },
        (options) => {
          expect(options.components).toHaveLength(3)
          expect(getCustomId(options.components![0])).toBe('button-match')
          expect(getCustomId(options.components![1])).toBe('button-join')
          expect(getCustomId(options.components![2])).toBe('button-leave')
        }
      )
    )
    // join 10
    await onInteractionCreated(
      createButtonInteraction(
        'button-join',
        {
          channelId,
          guildId,
          user: {
            id: 'user10',
            username: 'username10',
          },
        },
        (options) => {
          expect(options.components).toHaveLength(2)
          expect(getCustomId(options.components![0])).toBe('button-match')
          expect(getCustomId(options.components![1])).toBe('button-leave')
          expect(options).toMatchSnapshot()
        }
      )
    )

    // break
    await onInteractionCreated(
      createChatInputInteraction(
        'sr-break',
        {
          channelId,
          guildId,
          user: {
            id: 'user1',
            username: 'username1',
          },
        },
        (options) => {
          expect(options).toMatchSnapshot()
        }
      )
    )

    const roomCount = await prisma.room.count({
      where: {
        creatorUserId: 'user1',
        discordChannelId: channelId,
      },
    })
    expect(roomCount).toBe(0)
  })
})
