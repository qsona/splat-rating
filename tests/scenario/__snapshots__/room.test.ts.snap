// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`room scenario room scenario 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "button-join",
          "emoji": undefined,
          "label": "参加",
          "style": 1,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": "ゲーム募集 ガチエリア ホスト: username1 (R2000) @7~9",
}
`;

exports[`room scenario room scenario 2`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "button-join",
          "emoji": undefined,
          "label": "参加",
          "style": 1,
          "type": 2,
        },
      ],
      "type": 1,
    },
    {
      "components": [
        {
          "custom_id": "button-leave",
          "emoji": undefined,
          "label": "抜ける",
          "style": 2,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": "username2 さんがゲームに参加しました。 (R2100)
@6~8",
}
`;

exports[`room scenario room scenario 3`] = `
{
  "content": "username2 さんがゲームから抜けました。
@7~9",
}
`;

exports[`room scenario room scenario 4`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "button-match",
          "emoji": undefined,
          "label": "チーム分け",
          "style": 1,
          "type": 2,
        },
      ],
      "type": 1,
    },
    {
      "components": [
        {
          "custom_id": "button-join",
          "emoji": undefined,
          "label": "参加",
          "style": 1,
          "type": 2,
        },
      ],
      "type": 1,
    },
    {
      "components": [
        {
          "custom_id": "button-leave",
          "emoji": undefined,
          "label": "抜ける",
          "style": 2,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": "username8 さんがゲームに参加しました。 (R2700)
@0~2",
}
`;

exports[`room scenario room scenario 5`] = `
{
  "content": "username2 さんがゲームから抜けました。
@1~3",
}
`;

exports[`room scenario room scenario 6`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "button-match",
          "emoji": undefined,
          "label": "チーム分け",
          "style": 1,
          "type": 2,
        },
      ],
      "type": 1,
    },
    {
      "components": [
        {
          "custom_id": "button-leave",
          "emoji": undefined,
          "label": "抜ける",
          "style": 2,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": "username10 さんがゲームに参加しました。 (R2900)
@0~0",
}
`;

exports[`room scenario room scenario 7`] = `
{
  "content": "このチャンネルのゲームを解散しました。",
}
`;
